File: Java-Core/src/main/java/presto/type/TextType.java
package presto.type;

import java.util.Comparator;

import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.grammar.Identifier;
import presto.runtime.Context;
import presto.value.ICollection;
import presto.value.IValue;
import presto.value.Text;
import presto.value.ListValue;


public class TextType extends NativeType {

	static TextType instance = new TextType();
	
	public static TextType instance() {
		return instance;
	}
	
	private TextType() {
		super("Text");
	}
	
	@Override
	public Class<?> toJavaClass() {
		return String.class;
	}
	

	
	@Override
	public boolean isAssignableTo(Context context, IType other) {
		return (other instanceof TextType) || (other instanceof AnyType);
	}

	@Override
	public IType checkAdd(Context context, IType other, boolean tryReverse) throws SyntaxError {
		if(tryReverse)
			return this; // we're lhs, ok
		else
			return super.checkAdd(context, other, tryReverse);
	}
	
	@Override
	public IType checkMultiply(Context context, IType other, boolean tryReverse) throws SyntaxError {
		if(other instanceof IntegerType)
			return this;
		else
			return super.checkMultiply(context, other, tryReverse);
	}
	
	@Override
	public IType checkCompare(Context context, IType other) throws SyntaxError {
		if(other instanceof TextType || other instanceof CharacterType)
			return BooleanType.instance();
		return super.checkCompare(context, other);
	}
	
	@Override
	public IType checkItem(Context context, IType other) throws SyntaxError {
		if(other==IntegerType.instance())
			return CharacterType.instance();
		else
			return super.checkItem(context,other);
	}
	
	@Override
	public IType checkMember(Context context, Identifier id) throws SyntaxError {
		String name = id.toString();
       if ("length".equals(name))
            return IntegerType.instance();
       else
    	   return super.checkMember(context, id);
	}

	
	@Override
	public IType checkContains(Context context, IType other) throws SyntaxError {
		if(other instanceof TextType || other instanceof CharacterType)
			return BooleanType.instance();
		return super.checkContains(context, other);
	}
	
	@Override
	public IType checkContainsAllOrAny(Context context, IType other) throws SyntaxError {
		return BooleanType.instance();
	}
	
	@Override
	public IType checkSlice(Context context) throws SyntaxError {
		return this;
	}

	@Override
	public ListValue sort(Context context, ICollection<IValue> list) throws PrestoError {
		return this.<IValue>doSort(context,list, new Comparator<IValue>() {
			@Override
			public int compare(IValue o1, IValue o2) {
				return o1.toString().compareTo(o2.toString());
			};
		});
	}
	
	@Override
	public IValue convertJavaValueToPrestoValue(Object value) {
        if (value instanceof String)
            return new Text((String)value);
        else
            return (IValue)value; // TODO for now
	}
	
}

File: Java-CodeServer/src/main/resources/js/prompto/type/TextType.js
var NativeType = require("./NativeType").NativeType;
var CharacterType = null;
var IntegerType = require("./IntegerType").IntegerType;
var BooleanType = require("./BooleanType").BooleanType;
var AnyType = require("./AnyType").AnyType;
var Text = null; // circular dependency

exports.resolve = function() {
	CharacterType = require("./CharacterType").CharacterType;
	Text = require("../value/Text").Text;
}

function TextType()  {
	NativeType.call(this, "Text");
	return this;
}

TextType.prototype = Object.create(NativeType.prototype);
TextType.prototype.constructor = TextType;

TextType.instance = new TextType();

/*
	@Override
	public Class<?> toJavaClass() {
		return String.class;
	}
	
*/

TextType.prototype.isAssignableTo = function(context, other) {
	return (other instanceof TextType) || (other instanceof AnyType);
};

TextType.prototype.checkAdd = function(context, other, tryReverse) {
	// can add anything to text
	return this;
};

TextType.prototype.checkMultiply = function(context, other, tryReverse) {
	if(other instanceof IntegerType) {
		return TextType.instance;
	}
	return NativeType.prototype.checkMultiply.call(this, context, other, tryReverse);
};

TextType.prototype.checkCompare = function(context, other) {
	if(other instanceof TextType || other instanceof CharacterType) {
		return BooleanType.instance;
	}
	return NativeType.prototype.checkCompare.call(this, context, other);
};

TextType.prototype.checkItem = function(context, other) {
	if(other==IntegerType.instance) {
		return CharacterType.instance;
	} else {
		return NativeType.prototype.checkItem.call(this, context, other);
	}
};

TextType.prototype.checkMember = function(context, name) {
   if ("length"==name) {
	   return IntegerType.instance;
   } else {
	   return NativeType.prototype.checkMember.call(this, context, name);
   }
};

TextType.prototype.checkContains = function(context, other) {
	if(other instanceof TextType || other instanceof CharacterType) {
		return BooleanType.instance;
	}
	return NativeType.prototype.checkContains.call(this, context, other);
};

TextType.prototype.checkContainsAllOrAny = function(context, other) {
	return BooleanType.instance;
};

TextType.prototype.checkSlice = function(context) {
	return this;
};

TextType.prototype.sort = function(context, list) {

	function cmp(o1, o2) {
		o1 = o1.toString();
		o2 = o2.toString();
		return o1 > o2 ? 1 : o1 == o2 ? 0 : -1;
	}
	return NativeType.prototype.sort(context,list, cmp);
};

TextType.prototype.convertJavaScriptValueToPrestoValue = function(context, value, returnType) {
	if (typeof(value) == 'string') {
		return new Text(value);
	} else {
		return value; // TODO for now
	}
}

exports.TextType = TextType;
